### PR TITLE
AL-24 Add migrate alerts endpoint

### PR DESCRIPTION
**Pull request by jbowdidge-moj**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsalertsapi/resource/MigrateAlertsController.kt [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        content = [Content(schema = Schema(implementation = ErrorResponse::class))],
+      ),
+      ApiResponse(
+        responseCode = "409",
+        description = "Conflict, the person already has an active alert using the supplied alert code",
+        content = [Content(schema = Schema(implementation = ErrorResponse::class))],
+      ),
+    ],
+  )
+  @PreAuthorize("hasAnyRole('$ROLE_ALERTS_WRITER', '$ROLE_ALERTS_ADMIN', '$UPDATE_ALERT', '$ROLE_NOMIS_ALERTS')")
```
Comment (iterative-it) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsalertsapi/resource/MigrateAlertsController.kt:64]: The only roles that should be able to call this endpoint are ROLE_ALERTS_ADMIN and ROLE_NOMIS_ALERTS

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsalertsapi/resource/MigrateAlertIntTest.kt [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+
+    assertThat(alertEntity.migratedAt).isCloseToUtcNow(within(3, ChronoUnit.SECONDS))
+  }
+
+  @Test
+  fun `should not publish alert created event`() {
+    val request = migrateAlertRequest()
+
+    webTestClient.migrateAlert(request = request)
+    verify(hmppsQueueService, never()).findByTopicId(any())
```
Comment (iterative-it) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsalertsapi/resource/MigrateAlertIntTest.kt:203]: Unfortunately I think you'll have to put an artificial Thread.Sleep(1000) to be confident of this test assertion I think

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsalertsapi/service/MigrateAlertServiceTest.kt [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    underTest.migrateAlert(migrateAlertRequest)
+    verify(alertRepository).saveAndFlush(argumentCaptor.capture())
+    with(argumentCaptor.firstValue) {
+      assertThat(alertUuid).isNotNull()
+      assertThat(auditEvents()).hasSize(1)
+      assertThat(alertCode).isEqualTo(alertCodeRefusingToShieldInactive())
+      assertThat(migratedAt).isCloseToUtcNow(within(3, SECONDS))
+    }
+  }
+}
```
Comment (iterative-it) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsalertsapi/service/MigrateAlertServiceTest.kt:79]: Possibly worth adding both a happy path test here and one that asserts that `alert.publishedDomainEvents` is empty i.e. no events were registered

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsalertsapi/resource/MigrateAlertsController.kt (hunk 1 above)
Comment (jbowdidge-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsalertsapi/resource/MigrateAlertsController.kt:64]: Yes - I was going to check that with you.

**Review by iterative-it — APPROVED** [review]
LGTM

**Pull request merged**